File: kmp-mobile/android/src/main/kotlin/br/church/paz/android/ui/features/academy/CertificatesViewModel.kt
package br.church.paz.android.ui.features.academy

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.church.paz.shared.domain.repository.CourseCertificateEntry
import br.church.paz.shared.domain.repository.CourseRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class CertificatesUiState(
    val isLoading: Boolean = true,
    val certificates: List<CourseCertificateEntry> = emptyList(),
    val error: String? = null,
)

class CertificatesViewModel(
    private val courseRepository: CourseRepository,
) : ViewModel() {
    private val _uiState = MutableStateFlow(CertificatesUiState())
    val uiState: StateFlow<CertificatesUiState> = _uiState.asStateFlow()

    init {
        load()
    }

    fun load() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            runCatching { courseRepository.listCertificates() }
                .onSuccess { certificates -> _uiState.update { it.copy(isLoading = false, certificates = certificates) } }
                .onFailure { e -> _uiState.update { it.copy(isLoading = false, error = e.message) } }
        }
    }
}

File: kmp-mobile/ios/PazChurch/Features/Academy/CertificatesViewModel.swift
import Observation
import Shared
import SwiftUI

@MainActor
@Observable
class CertificatesViewModel {
    var isLoading = true
    var certificates: [CourseCertificateEntry] = []
    var error: String?

    private let courseRepository: CourseRepository

    init(courseRepository: CourseRepository) {
        self.courseRepository = courseRepository
    }

    func load() async {
        isLoading = true
        error = nil
        do {
            certificates = try await courseRepository.listCertificates()
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }
}
